### PR TITLE
Improve DependencyBuilder performance.

### PR DESCRIPTION
**Pull request by skorulis-ap**
This makes the DependencyBuilder roughly 3x faster in real world cases.

### CODE REVIEW

File: Sources/Knit/Module/DependencyTree.swift [hunk 1]
```diff
@@ -103,18 +102,20 @@ public struct DependencyTree: CustomDebugStringConvertible {
...
 
 /// Wrapper for the types to allow them to be hashable
-private struct AssemblyReference: Hashable {
+internal struct AssemblyReference: Hashable {
 
     let type: any ModuleAssembly.Type
-    let name: String
+    let id: ObjectIdentifier
+
+    var name: String { String(describing: type) }
```
Comment (skorulis-ap) [inline Sources/Knit/Module/DependencyTree.swift:110]: `String(describing: type)` was much slower than `ObjectIdentifier(type)`. This call is only for debugging.

File: Sources/Knit/Module/DependencyBuilder.swift [hunk 2]
```diff
@@ -212,3 +206,37 @@ extension ModuleAssembly {
     }
 
 }
+
+internal extension DependencyBuilder {
+
+    struct RegistrationCache {
```
Comment (skorulis-ap) [inline Sources/Knit/Module/DependencyBuilder.swift:212]: This cache allows more checks use .contains rather than iterating arrays for matches.

Comment (bradfol) [reply Sources/Knit/Module/DependencyBuilder.swift:212]: Nice 👍

File: Sources/Knit/Module/DependencyTree.swift (hunk 1 above)
Comment (bradfol) [reply Sources/Knit/Module/DependencyTree.swift:110]: Nit: `debugDescription` instead of `name`?

**Review by bradfol — APPROVED** [review]
Awesome, thanks!

**Pull request merged**